src/werkzeug/formparser.py

```diff
@@ -423,156 +411,322 @@ def validate_boundary(self, boundary):
-                        self.fail("could not decode transfer encoded chunk")
-
-                # we have something in the buffer from the last iteration.
-                # this is usually a newline delimiter.
-                if buf:
-                    yield _cont, buf
-                    buf = b""
-
-                # If the line ends with windows CRLF we write everything except
-                # the last two bytes.  In all other cases however we write
-                # everything except the last byte.  If it was a newline, that's
-                # fine, otherwise it does not matter because we will write it
-                # the next iteration.  this ensures we do not write the
-                # final newline into the stream.  That way we do not have to
-                # truncate the stream.  However we do have to make sure that
-                # if something else than a newline is in there we write it
-                # out.
-                if line[-2:] == b"\r\n":
-                    buf = b"\r\n"
-                    cutoff = -2
+        def _state_pre_term(self, line):
+            """State for the very beginning of the stream, before any content.
+            Eats empty lines until it finds a boundary line."""
+            if not line:
+                raise ValueError("Unexpected end of file")
+                return self._state_pre_term
+            line = line.rstrip(b"\r\n")
+            if not line:
+                return self._state_pre_term
+            if line == self._last_part:
+                return self._state_done
+            elif line == self._next_part:
+                self._headers = []
+                return self._state_headers
+            raise ValueError("Expected boundary at start of multipart data")
+
+        def _state_headers(self, line):
+            """State for field headers. They are parsed and left in
+            ``self._headers``."""
+            if line is None:
+                raise ValueError("Unexpected end of file during headers")
+            line = to_native(line)
+            line, line_terminated = _line_parse(line)
+            if not line_terminated:
+                raise ValueError("Unexpected end of line in multipart header")
+            if not line:
+                self._headers = Headers(self._headers)
+                return self._start_content()
+            if line[0] in " \t" and self._headers:
+                key, value = self._headers[-1]
+                self._headers[-1] = (key, value + "\n " + line[1:])
+            else:
+                parts = line.split(":", 1)
+                if len(parts) == 2:
+                    self._headers.append((parts[0].strip(), parts[1].strip()))
                 else:
-                    buf = line[-1:]
-                    cutoff = -1
-                yield _cont, line[:cutoff]
+                    raise ValueError("Malformed header")
+            return self._state_headers
+
+        def feed(self, lines):
```

Comment:
`self._output` only ever gets appended to, so this makes more sense as a generator. This also reduces some of the  confusion that arises from all of the state management happening in helper functions. A generator is also a better API for these `feed` functions generally.